File: iterator/golang/dinner.go
package main

// Dinner is the Aggregate interface
type Dinner interface {
	CreateDinner() Eating
}

// WeekendDinner is the concrete aggregate
type WeekendDinner struct {
	dishes []*Dish
}

func (w *WeekendDinner) CreateDinner() Eating {
	return &RestaurantEating{
		dishes: w.dishes,
	}
}

func (w *WeekendDinner) AddDish(name string) {
	w.dishes = append(w.dishes, &Dish{Name: name})
}

File: iterator/rust/iterator/src/dinner.rs
use crate::eating::{Dish, Eating};
use crate::restaurant::RestaurantEating;

pub trait Dinner {
    fn create_dinner(&self) -> Box<dyn Eating>;
}

pub struct WeekendDinner;

impl Dinner for WeekendDinner {
    fn create_dinner(&self) -> Box<dyn Eating> {
        let menu = vec![
            Dish { name: "Rice".to_string() },
            Dish { name: "Red Wine".to_string() },
        ];
        Box::new(RestaurantEating::new(menu))
    }
}
